resource: Fix race with event retries of deleted and subsequently recreated objects

joamaki opened the pull request:
See commits for detailed description.

```release-note
resource: Fix race condition in handling of Kubernetes object delete event retrying. In the very rare case when an object was created, deleted and re-created with the same name and the handling of the first deletion failed, the handling of delete event may have been retried even though the object was re-created. Only affected features using the Resource-library (LB IPAM, Mutual Auth and ClusterMesh).
```

joamaki committed:
resource: Add test for repeated deletions

Resource[T] does not correctly handle the events:
  Upsert -> Delete with Done(not-nil) -> Upsert (recreate)
    -> Delete (retry)

The retried delete event carries the old initial version of the
object causing the recreated object to be incorrectly deleted.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
resource: Fix double upserts on subscribe and retrying of delete events

Fix double upserts that were caused by store being manipulated without
synchronization with the subscriber queues by processing the deltas
under the resource read-lock and doing the initial key listing for
new subscriber with a write-lock. This way we cannot accidentally
see a key in the store and process it just before the key is queued.

As shown by test case in previous commit, the delete events are retried
with an old incorrect version of the object causing a recreated object
to be deleted.

Fix the deletion retrying by always queueing upserts and deletes by key
and keeping the last known state of objects emitted to the subscriber.
Only emit a delete event if the subscriber has seen its creation and
only use a version of the object that the subscriber has observed.

Fixes: 4101e2c768 ("k8s: Add resource package")
Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd commented on pkg/k8s/resource/resource_test.go:
```diff
@@ -256,6 +258,112 @@ func TestResource_WithFakeClient(t *testing.T) {
+		if ev.Kind == resource.Delete {
+			// Check that we don't go back in time
+			version, _ := strconv.ParseUint(ev.Object.ObjectMeta.ResourceVersion, 10, 64)
+			require.LessOrEqual(t, lastVersion, version, "expected always increasing ResourceVersion")
```
`LessOrEqual` vs `always increasing` - is equal ok or is it not? I read always increasing as also not constant, which would be allowed by the check

joamaki replied:
Must be `LessOrEqual` due to the retries.

bimmlerd commented on pkg/k8s/resource/resource_test.go:
```diff
@@ -256,6 +258,112 @@ func TestResource_WithFakeClient(t *testing.T) {
+				ev.Done(nil)
+			}
+		} else {
+			lastNode = ev.Object
```
probably deserves a `require non-nil check` here already?

bimmlerd commented on pkg/k8s/resource/resource.go:
```diff
@@ -483,21 +491,27 @@ func (r *resource[T]) Events(ctx context.Context, opts ...EventsOpt) <-chan Even
-				if err != nil || !exists {
-					event.Done(nil)
-					continue loop
+				if !exists || err != nil {
```
I think this is still racy.

Imagine the sequence `create (C) delete (D) create (C) delete (D)` for the same resource coming from k8s. The end state should be either four, two or no events emitted (CDCD, CD or nothing), but never starting with a D (deletion of not created resource), or ending with C (end state wrong, created instead of deleted). 
However, I claim that the current code can result in `DD`, i.e. double deletion of a resource we have not emitted a create for. The key queue coalesces `upsertOrDeleteEntry{"name"}` as long as we don't call `Get`. Hence, I can choose to let it coalesce the `CD` to just one, with which we enter the loop. Before we get to `store.GetByKey`, however, the third and fourth action `CD` have hit the store, which now does not contains the resource, and `deletedObjs` does. Hence, we emit just `D`, deleting a resource we didn't create.
Since the object in the `deletedObj` doesn't match, in the call to event.Done we do not remove it. The next call to `kq.Get()` thus relivers the `upsertOrDeleteEntry{"name"}` coalesced from the third and fourth action, and we emit another `D` event, since the store doesn't carry the resource, and the deletedObjs still do.

bimmlerd replied:
I've thought about this some more. I don't think there can be a sequence of k8s events in which the end state is 'deleted' in which the resource emits a sequence which ends in 'created'. Formulated differently, the race described above does not exist in the inverted (and plausibly worse) incarnation of missing the deletion of an object.

I think it's invariant that we emit creation events if and only if (iff) the store contains an object for a key. It's a guarantee of the workqueue that it will not coalesce into events which were observed (via `Get`). Further, for the sequence of events emitted by the resource to end with `C`, but the sequence of k8s events to end with `D`, the final `D` event must somehow be missed by the resource. The resource can only 'miss' events if they are coalesced in the key queue, but then we have the invariant that the store must have seen it (as it's the callback of the store to push it into the key queue). Hence, the store can't contain the object and also coalesce the event. Thus, the resource either observes the non-existence of the obj in the store, or gets the delete event. Therefore, the resource doesn't end a sequence of emitted events with `C` if it shouldn't. 

One step further, it seems an acceptable burden to place on the consumer to ignore (from their perspective) spurious deletes. Note that I think the resource shouldn't not emit (i.e. should emit) deletes for objects it hasn't emitted creates for, since consumers may have other (possibly indirect) sources of knowledge about an object and hence might know about it's creation even though this resource didn't provide an event.

joamaki replied:
Agreed on this. I'll look into tracking the emitted creations so deletions will always pair up with the create.

joamaki replied:
Fixed. Creations are now tracked and Resource[T] will only emit a delete event for objects that the subscriber has observed, e.g. if `create X`, `delete X` occurs in quick succession and the subscriber didn't see the `create X` it won't see the `delete X`.

bimmlerd reviewed (changes requested):
I think there's a race.

bimmlerd commented on pkg/k8s/resource/resource.go:
```diff
@@ -454,6 +454,10 @@ func (r *resource[T]) Events(ctx context.Context, opts ...EventsOpt) <-chan Even
 				debugInfo))
 		}
 
+		// The set of objects emitted to the subscriber. Used to avoid emitting a delete
```
Can't comment there, but I think the `Events` function doc comment should state the semantics around retries, given it's pretty subtle

joamaki replied:
Documented.

bimmlerd commented on pkg/k8s/resource/resource.go:
```diff
@@ -464,44 +468,74 @@ func (r *resource[T]) Events(ctx context.Context, opts ...EventsOpt) <-chan Even
+						event.Key = entry.key
+						event.Object = deletedObj.(T)
+					} else {
+						// Here we hit a race in which the object was removed from store
```
We could theoretically have also hit a case where `err != nil` from getting from the store (though I don't think the store can actually error). Nonetheless, if there was a persitent error, I think this would continually requeue the same event

joamaki replied:
Currently the store used here can't error since it's a simple in-memory hash map. I think the idea with having errors here is to be able to use e.g. a store that persists to disk. I'll change this to explicitly handle the unexpected error with a `Fatal`.

bimmlerd commented on pkg/k8s/resource/resource_test.go:
```diff
@@ -256,6 +258,110 @@ func TestResource_WithFakeClient(t *testing.T) {
+				ev.Done(nil)
+			}
+		} else if ev.Kind == resource.Upsert {
+			exists = true
```
Could also check the version not going backwards here, since we expect to observe the upserts in order

joamaki replied:
Added!

bimmlerd commented on pkg/k8s/resource/resource.go:
```diff
@@ -464,44 +468,74 @@ func (r *resource[T]) Events(ctx context.Context, opts ...EventsOpt) <-chan Even
+			var event Event[T]
 
 			switch entry := entry.(type) {
 			case syncEntry:
```
Kind of unrelated to the issue we're solving here, but should be documented at least: A consumer will receive a sync event regardless of whether the handling of the upserts was successful. That is, during the replay of the stuff in the store, if you fail to handle an event, you will receive the `Sync` before receiving the retried Upsert.

joamaki replied:
Good point. Added documentation about this to `Events()`.

bimmlerd reviewed (approved):
Couple of nits, but LGTM!

joamaki committed:
resource: Improve documentation on Events()

The semantics around retrying and the Sync event are subtle. Spell out the properties
in a comment for Events().

Signed-off-by: Jussi Maki <jussi@isovalent.com>

aanm commented on pkg/k8s/resource/resource.go:
```diff
@@ -454,6 +454,10 @@ func (r *resource[T]) Events(ctx context.Context, opts ...EventsOpt) <-chan Even
 
+		// The set of objects emitted to the subscriber. Used to avoid emitting a delete
+		// event for an object the subscriber has not seen.
+		var knownObjects sync.Map
```
it doesn't need to be sync.Map, does it? From a first glance it looks it's just one go routine accessing it?

joamaki replied:
It has to be a `sync.Map` (or map + mutex) since it's mutated from the `Done` function that may be called from anywhere by the subscriber. 

And honestly the only reason I really went with `sync.Map` here was to match with `keyQueue.deletedObjects` (which actually needed `sync.Map` for compare-and-swap) and because it was shorter. Could go with `sets.Set[Key]` + `Mutex` instead if it'd be preferred. Let me know what you think. I'll at the very least expand the comment to say that it's accessed concurrently.

aanm commented on pkg/k8s/resource/resource.go:
```diff
@@ -536,21 +570,28 @@ func (r *resource[T]) Events(ctx context.Context, opts ...EventsOpt) <-chan Even
 	// The entries must be added by value and not by pointer in order for
 	// them to be compared by value and not by pointer.
-	kq.Add(upsertEntry{key})
+	kq.Add(upsertOrDeleteEntry{key})
```
I would suggest to leave another comment in here explaining why an "Add" is passing an "upsertOrDelete" entry

joamaki replied:
`kq` is the workqueue, so we're adding an entry into the queue. It's not `Add` of an object (e.g. not the `T` in `Resource[T]` that is being upserted or deleted), but it's an add of a workqueue entry. Do you still want me to add a comment? What should the comment say?

Or what do you think about switching to `Enqueue` instead? It won't match the workqueue API it's wrapping, but it would be less confusing.

aanm replied:
> Or what do you think about switching to Enqueue instead? It won't match the workqueue API it's wrapping, but it would be less confusing.

Oh, I think this would definitely sound less confusing :)

joamaki replied:
oh wait, that's in client-go's workqueue, I can't change that. I'll make it less confusing by not embedding the `RateLimitingInterface`.

youngnick reviewed (approved):
LGTM with @aanm's changes.

bimmlerd commented on pkg/k8s/resource/resource.go:
```diff
@@ -522,100 +447,266 @@ func (r *resource[T]) Events(ctx context.Context, opts ...EventsOpt) <-chan Even
-	key Key
-	obj any
+				// In CI we detect if the objects were modified and panic
+				// this is a no-op in production environments.
```
Is this actually still true, or just copy pasta?

joamaki replied:
```
test/helpers/kubectl.go:                // "extraEnv[0].name":              "KUBE_CACHE_MUTATION_DETECTOR",
```

https://github.com/kubernetes/client-go/blob/v0.28.1/tools/cache/mutation_detector.go#L36

Heh, looks like we no longer set it anywhere, so no we don't check in CI.  And having it enabled by default is a fairly big hammer. I'll drop this from this fix as thinking about where and how to enable this is not trivial.

bimmlerd reviewed (approved):
Wow, nice simplification.

aanm reviewed (approved):
Can we adapt the pkg/k8s/informer.NewInformerWithStore for the pkg/k8s/resource use cases and use it?

Also, we should keep the mutator cache detector code, it was accidentally removed. I'm going to re-add it.
